lib/vm-agent.js

```diff
@@ -231,6 +239,52 @@ VmAgent.prototype.updateVmapiVm = function updateVmapiVm(vmUuid, callback) {
             var opts = {log: self.log, uuid: vmUuid, include_dni: true};
             var startLoad = (new Date()).getTime();
 
+            function _handleLoadErr(_vmUuid, err, next) {
+                var loadErr;
+
+                if (err && err.restCode === 'VmNotFound') {
+                    if (self.knownDniVms[_vmUuid]) {
+                        // DNI VM was deleted, we don't send this to VMAPI but
+                        // we do want to purge it from the list after some delay
+                        // in case other delete events are emitted, we still
+                        // need to know that this was DNI.
+                        if (!self.knownDniVms[_vmUuid].timer) {
+                            self.knownDniVms[_vmUuid].timer = setTimeout(
```

Comment:
I'm not sure if it's worth considering, but you might want to `unref()` that timer. Otherwise the libuv event loop will be held open for `DNI_PURGE_DELAY` even though nothing else would hold it open and it seems that purging DNI VMs is not critical in that case.